File: src/GrammaticOptimization/Production.java
/*
 * Production.java
 *
 * Created on June 3, 2007, 12:24 AM
 *
 * To change this template, choose Tools | Template Manager
 * and open the template in the editor.
 */

/**
 *
 * @author adilraja
 */
package GrammaticOptimization;

import java.util.*;
import java.util.ArrayList;
import java.util.Iterator;



public class Production extends ArrayList<Symbol>{
    
    
    private boolean recursive;
    private int minimumDepth;
    /**
     *Creates a new production with newLength elements
     */
    public Production(final int newLength){
        super();
        this.ensureCapacity(newLength);
        setRecursive(false);
        setMinimumDepth(32767>>1);
    }
    /**
     *Copy constructor; copy all symbols
     */
    public Production(final Production copy){
        Iterator<Symbol> itr = copy.iterator();
        while(itr.hasNext()) {
            Symbol tmpsymb=itr.next();
            if(tmpsymb.getType().toString().compareTo("TSymbol")==0){
                this.add(new Symbol(tmpsymb));// Terminal symbol, create new one
            }     
            else 
                this.add(tmpsymb);
        }
        setRecursive(copy.recursive);
        setMinimumDepth(copy.minimumDepth);
    }
    /**
     *Delete all terminal symbols and all symbol references
     */
    public void clearTerminals(){
        Iterator<Symbol> itr = this.iterator();
        while(itr.hasNext()) {
            Symbol tmpsymb=itr.next();
            if(tmpsymb.getType().toString().compareTo("TSymbol")==0){
                this.remove(tmpsymb);
            }            
            
        }
    }
    
    /**
     *Return the recursive nature of this production
     */
    
    public final boolean getRecursive(){
        return this.recursive;
    }
    
    /**
     *Update the recursive nature of this production
     */
    public void setRecursive(final boolean newRecursive){
        this.recursive=newRecursive;
    }
    /**
     *Return the minimum mapping depth of this production
     */
    public final int getMinimumDepth(){
        return this.minimumDepth;
    }
    
    
    /**
     *Update the minimum mapping depth of this production
     */
    public void setMinimumDepth(final int newMinimumDepth){
        this.minimumDepth=newMinimumDepth;
    }

    
    /**
     *Print the contents of this production
     */
    
    /**
     *This removes and adds a symbol at a particular location in this object. To be called latter from CFGrammar class.
     */
    public void insertSymbol(final Symbol symb, int pos){
        this.remove(pos);
        this.add(pos,symb);
    }
    
}
//file complete

File: libGEC++/libGE-0.26/src/Production.cpp
// Production.cpp -*- C++ -*-
#ifndef _PRODUCTION_CPP_
#define _PRODUCTION_CPP_

//#include<config.h>

#include<iostream>

#include"Production.h"

///////////////////////////////////////////////////////////////////////////////
// Creates a new Production with newLength elements.
Production::Production(const unsigned int newLength):vector<Symbol*>(){
#if (DEBUG_LEVEL >= 2)
	cerr << "'Production::Production(const unsigned int)' called\n";
#endif
	resize(newLength);
	setRecursive(false);
	setMinimumDepth(INT_MAX>>1);
}

///////////////////////////////////////////////////////////////////////////////
// Copy constructor; copy all symbols.
Production::Production(const Production &copy){
#if (DEBUG_LEVEL >= 2)
	cerr << "'Production::Production(const Production &copy)' called\n";
#endif
	const_iterator it=copy.begin();
	while(it!=copy.end()){
		if((*it)->getType()==TSymbol){
			// Terminal symbol, create new one
			push_back(new Symbol(**it));
		}
		else{
			push_back(*it);
		}
		it++;
	}
	setRecursive(copy.recursive);
	setMinimumDepth(copy.minimumDepth);
}

///////////////////////////////////////////////////////////////////////////////
// Destructor; call redefined operator clear().
Production::~Production(){
#if (DEBUG_LEVEL >= 2)
	cerr << "'Production::~Production()' called\n";
#endif
	clear();
}

///////////////////////////////////////////////////////////////////////////////
// Delete all terminal symbols and all symbol references.
void Production::clear(){
#if (DEBUG_LEVEL >= 2)
	cerr << "'void Production::clear()' called\n";
#endif
	iterator it=begin();
	while(it!=end()){
		if((*it)->getType()==TSymbol){
			// Terminal symbol, delete
		delete *it;
		}
		else{
		}
		it++;
	}
	erase(begin(),end());
}

///////////////////////////////////////////////////////////////////////////////
// Return the recursive nature of this production.
bool Production::getRecursive() const{
#if (DEBUG_LEVEL >= 2)
	cerr << "'bool Production::getRecursive() const' called\n";
#endif
	return recursive;
}

///////////////////////////////////////////////////////////////////////////////
// Update the recursive nature of this production.
void Production::setRecursive(const bool newRecursive){
#if (DEBUG_LEVEL >= 2)
	cerr << "'bool Production::setRecursive(const bool)' called\n";
#endif
	recursive=newRecursive;
}

///////////////////////////////////////////////////////////////////////////////
// Return the minimum mapping depth of this production.
unsigned int Production::getMinimumDepth() const{
#if (DEBUG_LEVEL >= 2)
	cerr << "'unsigned int Production::getMinimumDepth() const' called\n";
#endif
	return minimumDepth;
}

///////////////////////////////////////////////////////////////////////////////
// Update the minimum mapping depth of this production.
void Production::setMinimumDepth(const unsigned int newMinimumDepth){
#if (DEBUG_LEVEL >= 2)
	cerr << "'int Production::setMinimumDepth(const unsigned int)' called\n";
#endif
	minimumDepth=newMinimumDepth;
}

///////////////////////////////////////////////////////////////////////////////
// Pretty print the contents of the production.
ostream &operator<<(ostream &stream, const Production &p){
#if (DEBUG_LEVEL >= 2)
	cerr << "'ostream &operator<<(ostream &, const Production &)' called\n";
#endif
	string output;
	Production::const_iterator it=p.begin();
	while(it!=p.end()){
		output+=**it;
		it++;
	}
	return stream << output;
}

#endif

